src/posix/platform/multicast_routing.hpp

```diff
@@ -0,0 +1,191 @@
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef POSIX_PLATFORM_MULTICAST_ROUTING_HPP_
+#define POSIX_PLATFORM_MULTICAST_ROUTING_HPP_
+
+#include "openthread-posix-config.h"
+
+#if OPENTHREAD_CONFIG_BACKBONE_ROUTER_ENABLE
+
+#include <stdio.h>
+#include <stdlib.h>
+
+#include <chrono>
+#include <map>
+#include <set>
+#include <openthread/backbone_router_ftd.h>
+#include <openthread/openthread-system.h>
+
+#include "platform-posix.h"
+#include "core/common/non_copyable.hpp"
+#include "core/net/ip6_address.hpp"
+#include "lib/url/url.hpp"
+
+namespace ot {
+namespace Posix {
+
+/**
+ * This class implements Multicast Routing management.
+ *
+ */
+class MulticastRoutingManager : private NonCopyable
+{
+public:
+    /**
+     * This constructor initializes a Multicast Routing manager instance.
+     *
+     */
+    explicit MulticastRoutingManager()
+        : mMulticastRouterSock(-1)
+    {
+    }
+
+    /**
+     * This method initializes the Multicast Routing manager.
+     *
+     * @param[in]  aInstance  A pointer to an OpenThread instance.
+     *
+     */
+    void Init(otInstance *aInstance);
+
+    /**
+     * This method updates the fd_set and timeout for mainloop.
+     *
+     * @param[inout]    aReadFdSet      A reference to fd_set for polling read.
+     * @param[inout]    aMaxFd          A reference to the current max fd in fd_sets.
+     *
+     */
+    void UpdateFdSet(fd_set &aReadFdSet, int &aMaxFd) const;
+
+    /**
+     * This method performs Multicast Routing processing.
+     *
+     * @param[in]   aReadFdSet   A reference to read file descriptors.
+     *
+     */
+    void Process(const fd_set &aReadFdSet);
+
+    /**
+     * This method handles Thread state changes.
+     *
+     * @param[in] aInstance  A pointer to an OpenThread instance.
+     * @param[in] aFlags     Flags that denote the state change events.
+     *
+     */
+    void HandleStateChange(otInstance *aInstance, otChangedFlags aFlags);
+
+private:
+    enum
+    {
+        kMulticastForwardingCacheExpireTimeout    = 300, //< Expire timeout of Multicast Forwarding Cache (in seconds)
+        kMulticastForwardingCacheExpiringInterval = 60,  //< Expire interval of Multicast Forwarding Cache (in seconds)
+    };
+
+    enum MifIndex : uint8_t
+    {
+        kMifIndexNone     = 0xff,
```

Comment:
Is it possible to reuse otNetifIdentifier?

Reply:
It does not make absolute sense to use otNetifIdentifier. The MifIndex is a index to the MIF in the kernel. We used 0 for Thread and 1 for Backbone. These values are different from `otNetifIdentifier` values. 
Since they are different values, I don't feel the necessity to reuse otNetifIdentifier.

BTW, these values are private to `MulticastRoutingManager`.